lib/lists.nix

```diff
@@ -612,6 +616,42 @@ rec {
       if len < 2 then list
       else (sort strictLess pivot.left) ++  [ first ] ++  (sort strictLess pivot.right));
 
+  /*
+    Sort a list based on the default comparison of a derived property `b`.
+
+    This calls the function only once per element, so if your function is non-trivial,
+    it is easily twice as efficient as an unprepared [`sort`](#function-library-lib.lists.sort), and then a log(n) factor.
+
+    However, if the passed function is very cheap, `sort` may be cheaper,
+    because `sort` does not allocate an intermediate list of `(a, b)` pairs.
+
+    The items are returned in `b`-increasing order.
+
+    **Laws**:
+    ```nix
+    sortOn f == sort (a: b: f a < f b)
+    ```
+
+    Example:
+      sortOn (x: stringLength) [ "aa" "b" "cccc" ]
+      => [ "b" "aa" "cccc" ]
+
+    Type:
+      sortOn :: (a -> b) -> [a] -> [a], for comparable b
+  */
+  sortOn = f: list:
+    let
+      # Heterogenous list as pair may be ugly, but requires minimal allocations.
+      pairs = map (x: [(f x) x]) list;
+    in
+      map
+        (x: builtins.elemAt x 1)
+        (sort
+          # Compare the first element of the pairs
+          # Do not factor out the `<`, to avoid calls in hot code; duplicate instead.
+          (a: b: head a < head b)
+          pairs);
```

Comment:
It could be that this is more efficient:

```nix
let
  image = map f list;
in
map
  (elemAt list)
  (sort
    (i: j:
      elemAt image i < elemAt image j
    )
    (genList id (length list))
  )
```

Don't really mind it being the most efficient though :)